Iterator tutorials

BillWagner opened the pull request:
I've got questions in the PR message for exactly how to
manage this article in the main branch. Let's discuss those
as we review the content, and then I'll update with everything
ready to merge.

I marked this PR as WIP (work in progress) because writing it generated a few
questions about how we want to grow this section.

First, I believe the language tutorials will need to liberally cross-reference
each other. I'd like to have placeholders for where links should go (I've added
a couple in the above text.) However, that could discourage readers when the links
don't go to articles that don't yet exist.

Second, I'd like feedback on the tone of the article. I'm concerned that the middle
section was too close to a specification, and less of a tutorial. Thoughts?

Third, let's discuss the sample code. I tried to distill the code down to the smallest
possible samples to discuss the concept. That means they are less like a scenario, as
we would find in the other sections. I can update as we see fit.

Finally, I wanted to use a sub-directory for all C# tutorials. That broke
the TOC generator. I"m still ready about docfx, and if I get that right, I'll move
the files back. Also, I'm not sure why the TOC is not showing in the iterators article.

BillWagner committed:
WIP: Draft of iterators tutorial.

I've got questions in the PR message for exactly how to
manage this article in the main branch. Let's discuss those
as we review the content, and then I'll update with everything
ready to merge.

stevehoag commented:
For cross-referencing, I think we should go ahead and stub out any topics that we want to cross-reference. I believe there's boilerplate text to let readers no that they are just stubs for planned content, and it might actually inspire others to contribute.

The tone is generally on target, but I have a few concerns. The introduction could be simplified - we really just want to point out C#'s support for iterators, perhaps listing the relevant methods. Also - would someone coming from another language be familiar with the term "iterators", or do we need to explain.

It would also help to reorganize a bit. Provide the basic examples with simple explanations up front, then add a "deeper dive" section below with the implementation details. For example, your explanation of what the compiler creates is nice to know but isn't necessarily essential for someone wanting a quick intro to the language.

I think the sample code is fine - it illustrates the point without a lot of unnecessary overhead. Simple is good in this case.

I'll let someone else comment on the TOC issue - I don't have a clue.

Some specific edits:

Line 7 - enable, not enagle.
Line 17 - foreach
Line 93 - will the reader understand what a sealed type is? Maybe link to the Types topic?
Line 138 - instead of "I showed distinct `yield return` statements above...", use "The showed distinct `yield return` statements shown above..." 
Line 171 = use can't instead of cannot.

stevehoag commented:
I added my comments online. Not sure why notification mail wasn't sent automatically.


Sent from Outlook<http://aka.ms/weboutlook>

________________________________
From: Bill Wagner <notifications@github.com>
Sent: Wednesday, November 25, 2015 12:19:26 PM
To: dotnet/core-docs
Subject: [core-docs] WIP (Work in Progress): Iterator tutorials (#172)


I've got questions in the PR message for exactly how to
manage this article in the main branch. Let's discuss those
as we review the content, and then I'll update with everything
ready to merge.

I marked this PR as WIP (work in progress) because writing it generated a few
questions about how we want to grow this section.

First, I believe the language tutorials will need to liberally cross-reference
each other. I'd like to have placeholders for where links should go (I've added
a couple in the above text.) However, that could discourage readers when the links
don't go to articles that don't yet exist.

Second, I'd like feedback on the tone of the article. I'm concerned that the middle
section was too close to a specification, and less of a tutorial. Thoughts?

Third, let's discuss the sample code. I tried to distill the code down to the smallest
possible samples to discuss the concept. That means they are less like a scenario, as
we would find in the other sections. I can update as we see fit.

Finally, I wanted to use a sub-directory for all C# tutorials. That broke
the TOC generator. I"m still ready about docfx, and if I get that right, I'll move
the files back. Also, I'm not sure why the TOC is not showing in the iterators article.

________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/dotnet/core-docs/pull/172

Commit Summary

  *   WIP: Draft of iterators tutorial.

File Changes

  *   M docs/languages/c-sharp.md<https://github.com/dotnet/core-docs/pull/172/files#diff-0> (10)
  *   A docs/languages/csharp-iterators.md<https://github.com/dotnet/core-docs/pull/172/files#diff-1> (245)

Patch Links:

  *   https://github.com/dotnet/core-docs/pull/172.patch
  *   https://github.com/dotnet/core-docs/pull/172.diff

-
Reply to this email directly or view it on GitHub<https://github.com/dotnet/core-docs/pull/172>.

BillWagner committed:
First response to comments

I still need to rebase on PR #175, when approved, and add links to
appropriate topics.

BillWagner commented:
I updated the content based on @stevehoag 's comments. I still need to update links to other articles. I'll do that, and rebase once PR #175 is merged. I think that will make it much easier on everyone.

BethMassi commented on docs/languages/csharp-iterators.md:
```diff
@@ -0,0 +1,247 @@
+
+Both of these interfaces also have non-generic counterparts: `IEnumerable` and 
+`IEnumerator`. The generic versions are preferred for modern code.
+
```
Probably should link to the Generics topic here.

BethMassi commented on docs/languages/csharp-iterators.md:
```diff
@@ -0,0 +1,247 @@
+[lambda expressions](index.md) for more information. note: redirecting to the
+index seems the best way to handle a reference to a doc not yet written. I'm open
+to other suggestions.
+
```
That sounds like a reasonable approach.

BethMassi commented on docs/languages/csharp-iterators.md:
````diff
@@ -0,0 +1,247 @@
+
+```cs
+// C# versions 1 through 4:
+IEnumerator<int> enumerator = collection.GetEnumerator();
````
Should we bother with calling out syntax in earlier versions here considering our audience?

BillWagner replied:
I think we should, because the syntax is the same, but the compiler behavior is different. This was a breaking change, and there is content online that describes the old behavior. I thought that the only way to avoid any confusion would be to describe what it does not, and what the previous behavior was.  See: http://blogs.msdn.com/b/ericlippert/archive/2009/11/12/closing-over-the-loop-variable-considered-harmful.aspx and http://stackoverflow.com/questions/12112881/has-foreachs-use-of-variables-been-changed-in-c-sharp-5

BethMassi replied:
Ok sounds good. I wonder if we should consider that as a principle of the tutorial content here. i.e. If it's a breaking change then describe it, otherwise focus on only the current version. Thoughts?

BethMassi commented on docs/languages/csharp-iterators.md:
```diff
@@ -0,0 +1,247 @@
+handles all those nuances for you. The compiler will generate the correct code for
+any of these constructs. 
+
+
```
I'm thinking we should link to a full sample (that we create in the samples folder) and probably could pull in the sample as the reference code in this tutorial. Eventually, we want to get to the point where the samples are tested against the core builds. I'm not exactly sure what the md syntax is for pulling in lines of code though. /cc @cartermp

cartermp replied:
:+1: this is a good opportunity to start embedding buildable and runnable samples.  @blackdwarf or @chenkennt can you shed some light on how to embed source code (with a line range) from another file?

cartermp commented on docs/languages/csharp-iterators.md:
```diff
@@ -0,0 +1,247 @@
+# Introduction
```
I think this paragraph can be broken up to really bring home the value by enumerating (pun intended!) scenarios as a bulleted list:

* Doing something for each item in a collection
* Producing a way to enumerate a custom collection
* Extending LINQ or other collection libraries
* Writing a data pipeline which "streams" items efficiently through it

There's probably a few more, but these are areas I've seen the value add of iterators.

cartermp commented on docs/languages/csharp-iterators.md:
```diff
@@ -0,0 +1,247 @@
+features for both these scenarios. This article provides an overview
+of those features.
+
+# iterating with foreach
```
"i" should be capitalized.

cartermp commented on docs/languages/csharp-iterators.md:
```diff
@@ -0,0 +1,247 @@
+
+That's all there is to it. To iterate over all the contents of a collection,
+the `foreach` statement is all you need. The `foreach` statement isn't magic,
+though. It relies on two interfaces defined in the .NET core library in order
```
Perhaps a link or brief explanation about how `IEnumerable<T>` and `IEnumerator<T>` tie into this?  I like calling them out here, but without any info (however basic) about why these interfaces are important for iterating I would end up slightly confused.

cartermp commented on docs/languages/csharp-iterators.md:
````diff
@@ -0,0 +1,247 @@
+```cs
+{
+    var enumerator = collection.GetEnumerator();
+    try {
````
Would like to see allman braces here and below.

cartermp commented on docs/languages/csharp-iterators.md:
```diff
@@ -0,0 +1,247 @@
+Both of these interfaces also have non-generic counterparts: `IEnumerable` and 
+`IEnumerator`. The generic versions are preferred for modern code.
+
+# Enumeration sources with iterator methods
```
I mentioned this above, but I think a slightly more advanced scenario could go well here.  Here's a very rough scenario (feel free to suggest a better one):

For example, say I had a collection of objects which I wanted to turn into a CSV that a user could download on-demand.

A first approach would serialize the list as a CSV and then write that CSV to the response of a web controller.  This would work well with a small collection of items, but what if I had a list of a few thousand objects?  What of these objects had large values in them as well?  Memory would get slammed by this!  (fun fact: I've run into an `OutOfMemoryException` with a similar scenario in the past)

Iterators to the rescue!  Using `yield return` to defer execution allows for an `object --> CSV row` pipeline that each item in the collection runs through.  This "streaming" of objects through the CSV-to-response pipeline will allow your code to scale without consuming too much memory when your collection of objects grows.

stevehoag replied:
IMHO, I'm not sure that an advanced scenario is appropriate here - the tutorials should be an introduction to the basics of C#, not a comprehensive coverage.


Sent from Outlook<http://aka.ms/weboutlook>

________________________________
From: Phillip Carter <notifications@github.com>
Sent: Tuesday, December 1, 2015 2:00:36 PM
To: dotnet/core-docs
Cc: Steve Hoag
Subject: Re: [core-docs] WIP (Work in Progress): Iterator tutorials (#172)


In docs/languages/csharp-iterators.md<https://github.com/dotnet/core-docs/pull/172#discussion_r46345915>:

> +foreach (var item in collection)
> +{
> +   Console.WriteLine(item.ToString());
> +}
> +```
> +
> +That's all there is to it. To iterate over all the contents of a collection,
> +the `foreach` statement is all you need. The `foreach` statement isn't magic,
> +though. It relies on two interfaces defined in the .NET core library in order
> +to generate the code necessary to iterate a collection: `IEnumerable<T>` and
> +`IEnumerator<T>`.
> +
> +Both of these interfaces also have non-generic counterparts: `IEnumerable` and
> +`IEnumerator`. The generic versions are preferred for modern code.
> +
> +# Enumeration sources with iterator methods


I mentioned this above, but I think a slightly more advanced scenario could go well here. Here's a very rough scenario (feel free to suggest a better one):

For example, say I had a collection of objects which I wanted to turn into a CSV that a user could download on-demand.

A first approach would serialize the list as a CSV and then write that CSV to the response of a web controller. This would work well with a small collection of items, but what if I had a list of a few thousand objects? What of these objects had large values in them as well? Memory would get slammed by this! (fun fact: I've run into an OutOfMemoryException with a similar scenario in the past)

Iterators to the rescue! Using yield return to defer execution allows for an object --> CSV row pipeline that each item in the collection runs through. This "streaming" of objects through the CSV-to-response pipeline will allow your code to scale without consuming too much memory when your collection of objects grows.

-
Reply to this email directly or view it on GitHub<https://github.com/dotnet/core-docs/pull/172/files#r46345915>.

cartermp commented:
Left some feedback.  I like the way this is going.  I think the overall tone is fine, and the more specification-style tone in the middle is also pretty welcome.  We definitely want to call out certain gotchas with this documentation (such as not allowing `yield return` and `return` in the same method body), so I like that it does that.

As for the sample code, @BethMassi called out that we'll want the sample code here to be buildable and runnable.  I wouldn't worry too much about getting the docfx syntax in place for now.  The linq samples we have right now should be a good guideline for how to structure the samples.

chenkennt commented:
@cartermp currently we only support include a whole code sample file using the following markdown extension:
[!code-cs[sample](../../samples/linq/csharp/aggregate/Aggregate-Sample-1.cs)]

-cs can be changed to other languages.

We're working on supporting line range and a tagging based code snippet in this month, should be available soon

BillWagner committed:
respond to review comments

All source text comments are addressed.

BillWagner committed:
Move files, and fix links

BillWagner committed:
Add sample code from this article

I'm not including the expansion of the foreach keyword.
I think that's the right answer, but open to suggestions.

BillWagner committed:
Update the header to be consistent with other articles.

cartermp commented:
@chenkennt thanks!  I'll keep my ears open for when line ranges are available.

BillWagner commented:
I've responded to all the comments, and updated the samples, and the content.
I believe this is ready.

BethMassi commented on docs/languages/csharp/iterators.md:
```diff
@@ -1,5 +1,279 @@
+[lambda expressions](index.md) for more information. note: redirecting to the
+index seems the best way to handle a reference to a doc not yet written. I'm open
+to other suggestions.
+
```
Should remove this "note:.."

BillWagner replied:
Done. Updated the link to a new 'lambda-expressions.md' topic.

BillWagner committed:
Fix link on reference to lambda expressions

Add a tutorial document for that topic. (Placeholder at this time).

cartermp commented:
I'd say it LGTM.

@BethMassi @BillWagner what do you think of adding the author after the title?  We've been doing that for a few articles (e.g. Async and .NET Primer).  I think it would be a good idea just so we can show off community involvement here even more.

BethMassi commented:
Agreed. I think having by lines is a great idea and will encourage more writing contributions. We should put that note in the contributing.md :smile:

BethMassi commented:
@BillWagner by the way the tutorial turned out AWESOME. Should we add an issue to track creating the sample for it?

BillWagner commented:
@BethMassi I checked in the sample code already. It's part of this PR now. I can augment it with more code, if we need to.

stevehoag commented:
+1


Sent from Outlook<http://aka.ms/weboutlook>

________________________________
From: Beth Massi <notifications@github.com>
Sent: Thursday, December 3, 2015 11:42:07 AM
To: dotnet/core-docs
Cc: Steve Hoag
Subject: Re: [core-docs] Iterator tutorials (#172)


Agreed. I think having by lines is a great idea and will encourage more writing contributions. We should put that note in the contributing.md [:smile:]

-
Reply to this email directly or view it on GitHub<https://github.com/dotnet/core-docs/pull/172#issuecomment-161758553>.

BethMassi commented:
Awesome. LGTM!
